### PR TITLE
Add `UploadFailure` scenario

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Introduce envelope state](https://tools.hmcts.net/jira/browse/RPE-573)

### Change description ###

Notes:
- integration tests cover pretty much all the additions of this PR
- include message in exception of upload failure
- state integration was done in #38 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 1]
```diff
@@ -108,10 +108,18 @@ private void processZipFile(CloudBlobContainer container, String zipFilename)
...
-            envelopeProcessor.markAsUploaded(envelope, container.getName(), zipFilename);
-
-            cloudBlockBlob.delete();
+            try {
+                documentProcessor.processPdfFiles(pdfFiles, envelope.getScannableItems());
+                envelopeProcessor.markAsUploaded(envelope, container.getName(), zipFilename);
+
+                cloudBlockBlob.delete();
+            } catch (Exception exception) {
+                // catching any exception originated from document processor
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java]: It would be nice if we could log envelope name along with exception just in case update to DB fails ?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailedStatus.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        assertThat(envelopeState)
+            .extracting("container", "zipFileName", "status")
+            .hasSameElementsAs(ImmutableList.of(testContainer.getName(), ZIP_FILE_NAME, DOC_UPLOAD_FAILURE));
+        assertThat(envelopeState.getId()).isNotNull();
+        assertThat(envelopeState.getReason()).isNotBlank();
+        assertThat(envelopeState.getEnvelope().getId()).isEqualTo(actualEnvelope.getId());
+    }
+
+    @Test
+    public void should_record_failure_of_upload_when_document_management_throws_exception() throws IOException {
```
Comment (nitinprabhu) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailedStatus.java]: Could we have a scenario where even DB update fails ?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailedStatus.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    private void uploadZipToBlobStore() throws Exception {
+        byte[] zipFile = toByteArray(getResource(ZIP_FILE_NAME));
+
+        CloudBlockBlob blockBlobReference = testContainer.getBlockBlobReference(ZIP_FILE_NAME);
+        blockBlobReference.uploadFromByteArray(zipFile, 0, zipFile.length);
+    }
+
+    private List<Pdf> getUploadResources() throws IOException {
```
Comment (nitinprabhu) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailedStatus.java]: I think method like these could be extracted to a common base class instead of repeating it in each class ? Could be a separate PR for refactoring though.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java (hunk 1 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java]: logging should come in together with exception handler up in another level. this additional "exception wrapping" is for recording states. left logging out of it intentionally

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailedStatus.java (hunk 3 above)
Comment (doncem) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailedStatus.java]: True. best to leave for later once all the work around it is done and you are able to see bigger picture. This function will also change in the next PR 😉
